### PR TITLE
Api manipulation method values

### PR DESCRIPTION
**Pull request by jonathanKingston**
**Asana Task/Github Issue:** https://app.asana.com/1/137249556945/task/1214898456412755?focus=true

## Description

This should allow remote configuration of value descriptors that are functions.

## Testing Steps

Set the following config:

```json
                    {
                        "condition": [
                            {
                                "domain": "claude.ai"
                            }
                        ],
                        "patchSettings": [
                            {
                                "op": "add",
                                "path": "/apiChanges/MediaDevices.prototype.getUserMedia",
                                "value": {
                                    "type": "descriptor",
                                    "getterValue": {
                                        "type": "function",
                                        "functionName": "debug"
                                    }
                                }
                            }
                        ]
                    },
```

image

## Checklist

*Please tick all that apply:*

- [ ] I have tested this change locally
- [ ] I have tested this change locally in all supported browsers
- [ ] This change will be visible to users
- [ ] I have added automated tests that cover this change
- [ ] I have ensured the change is gated by config
- [ ] This change was covered by a ship review
- [ ] This change was covered by a tech design
- [ ] Any dependent config has been merged

---

> [!NOTE]
> **Medium Risk**
> Expands DOM API override capabilities (including methods and setters) and changes descriptor wrapping/definition behavior, which could affect site compatibility or detection resistance if edge cases are missed. Risk is mitigated by extensive new unit tests and schema validation coverage.
> 
> **Overview**
> Enables remote `apiManipulation` config to override **value descriptors** (including methods) and **setter-only/accessor overrides** via new `value` and `setterValue` fields, with stricter validation to prevent mixing accessor/value descriptor shapes.
> 
> When overriding functions, replacements are now *masked* to preserve observable identity (`toString()`, `name`, `length`, including native-code appearance) and use captured globals (`Reflect.apply`, `Object.defineProperty`) to reduce page tampering. Inherited/prototype-chain properties can be overridden by shadow-defining an own property (without requiring `define: true`), using a new `mergePropertyDescriptors` helper shared by `wrapProperty`.
> 
> Updates the privacy-config dependency and adjusts an integration test config JSON shape; adds extensive unit tests for method replacement, inherited overrides, setter interception, and schema validation against updated privacy-configuration.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d2209bb7f4fc0f8ffae14e9768eb06a83d6b7b0b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by jonathanKingston** [comment]
@claude Can you validate that if an existing value function descriptor exists, that we still can remotely override it with a new function descriptor (aka not with define: true)?

**Comment by jonathanKingston** [comment]
@cursoragent  Can you validate that if an existing value function descriptor exists, that we still can remotely override it with a new function descriptor (aka not with define: true)?

File: injected/src/features/api-manipulation.js [hunk 1]
```diff
@@ -111,29 +115,66 @@ export default class ApiManipulation extends ContentFeature {
...
+     * @returns {Partial<import('../wrapper-utils.js').StrictPropertyDescriptor>}
+     */
+    createApiDescriptor(descriptorKind, configSetting, change) {
+        const descriptor =
+            descriptorKind === 'getter'
+                ? {
+                      get: () => processAttr(configSetting, undefined),
+                  }
+                : {
+                      value: processAttr(configSetting, undefined),
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js]: @cursoragent try and resolve. IF this explodes the changeset or is hard to validate; continue as is.

File: injected/src/features/api-manipulation.js [hunk 2]
```diff
@@ -111,29 +116,101 @@ export default class ApiManipulation extends ContentFeature {
...
+     * @param {Function} replacementFn - configured replacement to invoke
+     * @param {Function} origFn - original DOM method we are masking against
+     * @returns {Function}
+     */
+    maskMethodReplacement(replacementFn, origFn) {
+        const wrapper = function () {
+            return Reflect.apply(replacementFn, this, arguments);
+        };
+        Object.defineProperty(wrapper, 'name', { value: origFn.name, configurable: true });
+        Object.defineProperty(wrapper, 'length', { value: origFn.length, configurable: true });
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js]: @cursoragent fix

File: injected/src/features/api-manipulation.js [hunk 3]
```diff
@@ -111,29 +116,105 @@ export default class ApiManipulation extends ContentFeature {
...
+     * @param {'getter' | 'value'} descriptorKind
+     * @returns {import('../wrapper-utils.js').StrictPropertyDescriptor}
+     */
+    createDefineDescriptor(descriptor, descriptorKind) {
+        if (descriptorKind === 'value') {
+            const valueDescriptor = /** @type {{ value: any, enumerable?: boolean, configurable?: boolean }} */ (descriptor);
+            return {
+                value: valueDescriptor.value,
+                writable: true,
+                enumerable: typeof valueDescriptor.enumerable !== 'boolean' ? true : valueDescriptor.enumerable,
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js:280]: I think this is a pretty big edge case at this point.

**Review by laghee — APPROVED** [review]
We could add some logging in here in the future to prevent potential dev confusion (e.g., Cursor claims accidentally wrapping an instance instead of the prototype would silently fail, as do invalid API changes), but otherwise LGTM

File: injected/src/features/api-manipulation.js [hunk 4]
```diff
@@ -111,29 +122,136 @@ export default class ApiManipulation extends ContentFeature {
...
-                    ...descriptor,
-                    enumerable: typeof descriptor.enumerable !== 'boolean' ? true : descriptor.enumerable,
-                    configurable: typeof descriptor.configurable !== 'boolean' ? true : descriptor.configurable,
+            // `setterValue` is intentionally invoked through processAttr on every assignment so
+            // configurations using `functionMap.debug` log per-assignment. When omitted, we leave
+            // `set` unset on the new descriptor so wrapProperty's spread merge preserves the
+            // original setter (existing behaviour).
+            if (change.setterValue !== undefined) {
+                const setterSetting = change.setterValue;
+                descriptor.set = function setter(v) {
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js:252]: @cursoragent fix this

File: injected/src/features/api-manipulation.js [hunk 5]
```diff
@@ -111,29 +122,136 @@ export default class ApiManipulation extends ContentFeature {
...
+        // skips the `get` half.
+        if (!descriptorKind || (descriptorKind === 'value' && configSetting === undefined)) {
+            return;
+        }
+        const descriptor = this.createApiDescriptor(descriptorKind, configSetting, change);
+        // If `define` is true and the property is not already an OWN property of `api`, define it
+        // directly. This covers both the "property does not exist at all" case and the
+        // "property is inherited via the prototype chain" case (e.g. shadow-defining
+        // MediaDevices.prototype.addEventListener which is inherited from EventTarget.prototype).
+        if (change.define === true && !hasOwnProperty.call(api, key)) {
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js]: @cursoragent fix

File: injected/src/features/api-manipulation.js [hunk 6]
```diff
@@ -177,14 +190,30 @@ export default class ApiManipulation extends ContentFeature {
...
+            if (configSetting !== undefined) {
+                descriptor.get = () => processAttr(configSetting, undefined);
+            }
+            // `setterValue` is intentionally invoked through processAttr on every assignment so
+            // configurations using `functionMap.debug` log per-assignment. When omitted, we leave
+            // `set` unset on the new descriptor so wrapProperty's spread merge preserves the
+            // original setter (existing behaviour).
+            if (change.setterValue !== undefined) {
+                const setterSetting = change.setterValue;
+                descriptor.set = function setter(v) {
```
Comment (jonathanKingston) [inline injected/src/features/api-manipulation.js:252]: >Property 'set' does not exist on type 'Partial<StrictPropertyDescriptor>'.

@cursoragent  can you fix this too, fixing the type def.

File: injected/src/features/api-manipulation.js [hunk 7]
```diff
@@ -111,29 +122,136 @@ export default class ApiManipulation extends ContentFeature {
...
+        const value = change.value;
+        const descriptorKind = getterValue !== undefined || setterValue !== undefined ? 'getter' : value !== undefined ? 'value' : undefined;
+        const configSetting = descriptorKind === 'getter' ? getterValue : value;
+        // `setterValue` may be supplied on its own (override only the setter, keep the original
+        // getter); in that case configSetting (getterValue) is undefined and createApiDescriptor
+        // skips the `get` half.
+        if (!descriptorKind || (descriptorKind === 'value' && configSetting === undefined)) {
+            return;
+        }
+        const descriptor = this.createApiDescriptor(descriptorKind, configSetting, change);
```
Comment (jonathanKingston) [inline injected/src/features/api-manipulation.js:136]: > Argument of type 'ConfigSetting | undefined' is not assignable to parameter of type 'ConfigSetting | ConfigSetting[]'.

@cursoragent  we need to get this type switched up to support value, getter and setter now. Is that the issue with the ts fail?

File: injected/src/features/api-manipulation.js [hunk 8]
```diff
@@ -66,9 +75,11 @@ export default class ApiManipulation extends ContentFeature {
...
      * @property {import('../utils.js').ConfigSetting} [getterValue] - The value returned from a getter.
+     * @property {import('../utils.js').ConfigSetting} [setterValue] - The function invoked when the property is assigned. Used alongside (or instead of) getterValue to override accessor-style properties such as event handlers (e.g., `MediaDevices.prototype.ondevicechange`).
+     * @property {import('../utils.js').ConfigSetting} [value] - The value assigned to a value descriptor, including methods.
      * @property {boolean} [enumerable] - Whether the property is enumerable.
      * @property {boolean} [configurable] - Whether the property is configurable.
-     * @property {boolean} [define] - Whether to define the property if it does not exist.
+     * @property {boolean} [define] - Whether to define the property if it is not already an own property of `api`. When true and the property is missing or only inherited via the prototype chain, a new own property is defined (shadowing any inherited one). When false (default), the change is merged with any existing own descriptor; properties that exist only via the prototype chain are skipped.
```
Comment (jonathanKingston) [inline injected/src/features/api-manipulation.js]: @cursoragent Don't we just pass false here, and the code should work as expected? Or do we have to pass define? Can you investigate?

File: injected/src/features/api-manipulation.js [hunk 9]
```diff
@@ -111,29 +129,198 @@ export default class ApiManipulation extends ContentFeature {
...
+        const setterValue = change.setterValue;
+        const value = change.value;
+        const descriptorKind =
+            getterValue !== undefined || setterValue !== undefined ? 'getter' : value !== undefined ? 'value' : undefined;
+        const configSetting = descriptorKind === 'getter' ? getterValue : value;
+        // `setterValue` may be supplied on its own (override only the setter, keep the original
+        // getter); in that case configSetting (getterValue) is undefined and createApiDescriptor
+        // skips the `get` half.
+        if (!descriptorKind || (descriptorKind === 'value' && configSetting === undefined)) {
+            return;
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js:134]: @cursoragent I don't think we need this define:true branch. Can you validate?


```json
        "apiChanges": {
          "MediaDevices.prototype.addEventListener": {
            "type": "descriptor",
            "define": true,
            "value": {
              "type": "function",
              "functionName": "noop"
            }
          },
          "MediaDevices.prototype.ondevicechange": {
            "type": "descriptor",
            "getterValue": {
              "type": "undefined"
            },
            "setterValue": {
              "type": "function",
              "functionName": "noop"
            }
          },
          "MediaDevices.prototype.removeEventListener": {
            "type": "descriptor",
            "define": true,
            "value": {
              "type": "function",
              "functionName": "noop"
            }
          }
```
          
          Should be:
          @cursoragent I don't think we need this define:true branch. Can you validate?


```json
        "apiChanges": {
          "MediaDevices.prototype.addEventListener": {
            "type": "descriptor",
            "value": {
              "type": "function",
              "functionName": "noop"
            }
          },
          "MediaDevices.prototype.ondevicechange": {
            "type": "descriptor",
            "getterValue": {
              "type": "undefined"
            },
            "setterValue": {
              "type": "function",
              "functionName": "noop"
            }
          },
          "MediaDevices.prototype.removeEventListener": {
            "type": "descriptor",
            "value": {
              "type": "function",
              "functionName": "noop"
            }
          }
```
          
          Then it should 'just' work with the setter and value modifications?

[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js:134]: @cursoragent what I'm saying is I don't want define:true to work, it should bail out

File: injected/src/features/api-manipulation.js [hunk 10]
```diff
@@ -66,9 +82,11 @@ export default class ApiManipulation extends ContentFeature {
...
+     * @property {import('../utils.js').ConfigSetting} [setterValue] - The function invoked when the property is assigned. Used alongside (or instead of) getterValue to override accessor-style properties such as event handlers (e.g., `MediaDevices.prototype.ondevicechange`).
+     * @property {import('../utils.js').ConfigSetting} [value] - The value assigned to a value descriptor, including methods.
      * @property {boolean} [enumerable] - Whether the property is enumerable.
      * @property {boolean} [configurable] - Whether the property is configurable.
-     * @property {boolean} [define] - Whether to define the property if it does not exist.
+     * @property {boolean} [define] - Whether to define the property when it does not exist anywhere on `api` or its prototype chain. Typically not needed: by default, own descriptors are merged and inherited descriptors are automatically shadow-defined as an own property (e.g. `MediaDevices.prototype.addEventListener` inherited from `EventTarget.prototype`). Set to `true` only to create a brand-new property that is missing from both `api` and every prototype in the chain.
```
Comment (jonathanKingston) [inline injected/src/features/api-manipulation.js]: >      * @property {boolean} [define] - Whether to define the property when it does not exist anywhere on `api` or its prototype chain. Typically not needed: by default, own descriptors are merged and inherited descriptors are automatically shadow-defined as an own property (e.g. `MediaDevices.prototype.addEventListener` inherited from `EventTarget.prototype`). Set to `true` only to create a brand-new property that is missing from both `api` and every prototype in the chain.

@cursoragent I think we can change this comment back again now you've changed the define behaviour back.

File: injected/src/features/api-manipulation.js [hunk 11]
```diff
@@ -111,29 +122,177 @@ export default class ApiManipulation extends ContentFeature {
...
-                descriptor.configurable = change.configurable;
+            return;
+        }
+        // When replacing an existing method-valued descriptor, mask the replacement
+        // against the original method so `toString()`, `toString.toString()`, `.name`
+        // and `.length` continue to resemble the native method. Without this, sites can
+        // trivially detect the override via `fn.toString()` / `fn.name`.
+        if (descriptorKind === 'value') {
+            const valueDescriptor = /** @type {{ value?: any }} */ (descriptor);
+            if (typeof valueDescriptor.value === 'function' && typeof origDescriptor.value === 'function') {
```
Comment (jonathanKingston) [inline injected/src/features/api-manipulation.js:150]: >   The merge { ...origDescriptor, ...descriptor } must remain a single valid ECMAScript property descriptor (not both accessor
  fields and value/writable).
>
>  • If config uses value but findPropertyDescriptor finds an accessor (or the opposite), Object.defineProperty will throw.
>  • That can happen only with a bad remote config for a given scope key; the XOR check (getterValue/setterValue vs value)
    does not compare the chosen shape to the live property kind on the prototype.

@cursoragent  Should we check `origDescriptor` for the presence of set or get to prevent the throw? Does the same problem happen in inverse for setter/getter with an existing value.

Can you find a way to make this work cleanly across all call sites?

File: injected/src/features/api-manipulation.js [hunk 12]
```diff
@@ -111,29 +122,180 @@ export default class ApiManipulation extends ContentFeature {
...
+     * @returns {PropertyDescriptor | undefined}
+     */
+    findPropertyDescriptor(obj, key) {
+        let current = obj;
+        while (current) {
+            const descriptor = getOwnPropertyDescriptor(current, key);
+            if (descriptor) {
+                return descriptor;
+            }
+            current = Object.getPrototypeOf(current);
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply injected/src/features/api-manipulation.js:192]: Will do in a follow up.

**Review by GuiltyDolphin — APPROVED** [review]
Discussed in call:
- both suspicious of the shadow define, but willing to proceed
- if the YT/linkedin/claude changes cause any problems we can roll back via config
- if existing API overrides are affected then this is harder to roll back via config

Config change is merged so this should apply once Windows adopts new CSS version.

Once you've finished running the final validations re existing APIs and have no concerns, feel free to merge.
